metadata/src/utils/retain.rs

```diff
@@ -5,187 +5,134 @@
 //! Utility functions to generate a subset of the metadata.
 
 use crate::{
-    ExtrinsicMetadata, Metadata, OuterEnumsMetadata, PalletMetadataInner, RuntimeApiMetadataInner,
-    StorageEntryType,
+    ExtrinsicMetadata, Metadata, PalletMetadataInner, RuntimeApiMetadataInner, StorageEntryType,
+};
+use alloc::collections::{BTreeSet, VecDeque};
+use scale_info::{
+    PortableType, TypeDef, TypeDefArray, TypeDefBitSequence, TypeDefCompact, TypeDefComposite,
+    TypeDefSequence, TypeDefTuple, TypeDefVariant,
 };
-use alloc::collections::BTreeMap;
-use hashbrown::HashSet;
-use scale_info::TypeDef;
-
-/// Collect all type IDs needed to represent the provided pallet.
-fn collect_pallet_types(pallet: &PalletMetadataInner, type_ids: &mut HashSet<u32>) {
-    if let Some(storage) = &pallet.storage {
-        for entry in storage.entries() {
-            match entry.entry_type {
-                StorageEntryType::Plain(ty) => {
-                    type_ids.insert(ty);
-                }
-                StorageEntryType::Map {
-                    key_ty, value_ty, ..
-                } => {
-                    type_ids.insert(key_ty);
-                    type_ids.insert(value_ty);
-                }
-            }
-        }
-    }
 
-    if let Some(ty) = pallet.call_ty {
-        type_ids.insert(ty);
-    }
+#[derive(Clone)]
+struct TypeSet {
+    seen_ids: BTreeSet<u32>,
+    work_set: VecDeque<u32>,
+}
 
-    if let Some(ty) = pallet.event_ty {
-        type_ids.insert(ty);
+impl TypeSet {
+    fn new() -> Self {
+        Self {
+            seen_ids: BTreeSet::new(),
+            work_set: VecDeque::with_capacity(22),
```

Comment:
Why 22?

Reply:
No particular reason besides that the biggest struct I've seen in the wild had ~20 fields

Reply:
yeah, would be good with a comment because I will ask the same everytime I see that :D

Reply:
added a comment, also simplified a bit to keep workset size smaller, now its usually around 30-50 elements